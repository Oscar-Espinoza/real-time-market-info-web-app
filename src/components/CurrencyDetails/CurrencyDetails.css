.currency-detail {
  display: flex;
  flex-direction: column;
  align-items: center;
  font-family: Arial, sans-serif;
  position: relative;
}

.back-home-btn {
  position: absolute;
  left: 5px;
  top: 5px;
  font-size: 18px;
  color: white;
}

.currency-title {
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  gap: 10px;
  min-height: 30vh;
  width: 100%;
}

.title-img {
  font-size: 20vh;
  color: #b03967;
}

.title span {
  font-size: 15px;
  width: 100%;
  display: inline-block;
  text-align: end;
}

.stats-separator {
  font-weight: bold;
  width: 100%;
  padding: 5px 10px;
}

.currency-details-list {
  list-style: none;
  display: flex;
  width: 100%;
  flex-direction: column;
  height: auto;
  font-family: Lato;
  font-style: normal;
}

.currency-details-row {
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding: 30px 20px;
}

.currency-details-left {
  font-weight: bold;
}

.currency-details-right {
  font-weight: 600;
}

.positive-per {
  color: green;
}

.negative-per {
  color: red;
}
