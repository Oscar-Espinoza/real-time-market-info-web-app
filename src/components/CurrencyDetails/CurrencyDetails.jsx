/* eslint linebreak-style: ["error", "windows"] */
import React from 'react';
import { BsTriangleFill, BsCurrencyExchange, BsArrowLeftCircle } from 'react-icons/bs';
import { Link, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import './currencyDetails.css';

export default function CurrencyDetails(props) {
  const { currencyName } = useParams();
  const { views } = props;
  const currency = useSelector((state) => state.currencies.currencies
    .find((currency) => currency.name === currencyName));

  function nFormatter(num, digits) {
    const lookup = [
      { value: 1, symbol: '' },
      { value: 1e3, symbol: ' Thousand' },
      { value: 1e6, symbol: ' Million' },
      { value: 1e9, symbol: ' Billion' },
      { value: 1e12, symbol: 'Tril' },
      { value: 1e15, symbol: 'P' },
      { value: 1e18, symbol: 'E' },
    ];
    const rx = /\.0+$|(\.[0-9]*[1-9])0+$/;
    const item = lookup.slice().reverse().find((item) => num >= item.value);
    return item ? (num / item.value).toFixed(digits).replace(rx, '$1') + item.symbol : '0';
  }

  return (
    <div className="currency-detail">
      <Link to="/" className="back-home-btn">
        <BsArrowLeftCircle />
      </Link>
      <div className="currency-title">
        <div className="title-img">
          <BsCurrencyExchange />
        </div>
        <h2 className="title">
          {currency.name || ''}
          {' '}
          (
          {currency.symbol || ''}
          )
          {' '}
          <br />
          <span>
            {views}
            {' '}
            Views
          </span>
        </h2>
      </div>

      <div className="stats-separator dark-bg">
        Stats for
        {' '}
        {currency.name || ''}
      </div>

      <div className="currency-details-list">
        <div className="currency-details-row light-bg">
          <div className="currency-details-left">Rank:</div>
          <div className="currency-details-right">{currency.rank || ''}</div>
        </div>
        <div className="currency-details-row dark-bg">
          <div className="currency-details-left">Supply:</div>
          <div className="currency-details-right">{nFormatter(currency.supply || '')}</div>
        </div>
        <div className="currency-details-row light-bg">
          <div className="currency-details-left">Max Supply:</div>
          <div className="currency-details-right">{currency.maxSupply ? nFormatter(currency.maxSupply) : 'No defined cap'}</div>
        </div>
        <div className="currency-details-row dark-bg">
          <div className="currency-details-left">Market Cap (USD):</div>
          <div className="currency-details-right">{nFormatter(currency.marketCapUsd) || ''}</div>
        </div>
        <div className="currency-details-row light-bg">
          <div className="currency-details-left">Change (24hr):</div>
          <div className="currency-details-right">
            {currency.changePercent24Hr[0] === '-'
              ? currency.changePercent24Hr.slice(0, 5)
              : `+ ${currency.changePercent24Hr.slice(0, 4)}`}
            %
            {' '}
            <BsTriangleFill
              className={currency.changePercent24Hr[0] !== '-'
                ? 'positive-per'
                : 'negative-per'}
            />
          </div>
        </div>
      </div>
    </div>
  );
}

CurrencyDetails.propTypes = {
  views: PropTypes.number,
};

CurrencyDetails.defaultProps = {
  views: Math.floor(Math.random() * 6000) + 1000,
};
