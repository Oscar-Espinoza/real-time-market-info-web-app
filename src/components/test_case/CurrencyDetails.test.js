import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import CurrencyDetails from '../CurrencyDetails/CurrencyDetails';

require('@testing-library/jest-dom');

const mockStore = configureMockStore();

describe('CurrencyDetails component', () => {
  it('renders currency details without any error', () => {
    const initialState = {
      currencies: {
        currencies: [
          {
            id: 'bitcoin',
            rank: '1',
            symbol: 'BTC',
            name: 'bitcoin',
            supply: '17193925.0000000000000000',
            maxSupply: '21000000.0000000000000000',
            marketCapUsd: '119179791817.6740161068269075',
            volumeUsd24Hr: '2928356777.6066665425687196',
            priceUsd: '6931.5058555666618359',
            changePercent24Hr: '-0.8101417214350335',
            vwap24Hr: '7175.0663247679233209',
          },
        ],
      },
    };
    const store = mockStore(initialState);

    const output = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/bitcoin']}>
          <Routes>
            <Route path="/:currencyName" element={<CurrencyDetails views={1233} />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
    expect(output).toMatchSnapshot();
  });
});
