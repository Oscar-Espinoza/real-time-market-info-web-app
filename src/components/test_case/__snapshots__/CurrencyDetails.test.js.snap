// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyDetails component renders currency details without any error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="currency-detail"
      >
        <a
          class="back-home-btn"
          href="/"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zm-4.5-.5a.5.5 0 0 1 0 1H5.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L5.707 7.5H11.5z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
        <div
          class="currency-title"
        >
          <div
            class="title-img"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 5a5.002 5.002 0 0 0 4.027 4.905 6.46 6.46 0 0 1 .544-2.073C3.695 7.536 3.132 6.864 3 5.91h-.5v-.426h.466V5.05c0-.046 0-.093.004-.135H2.5v-.427h.511C3.236 3.24 4.213 2.5 5.681 2.5c.316 0 .59.031.819.085v.733a3.46 3.46 0 0 0-.815-.082c-.919 0-1.538.466-1.734 1.252h1.917v.427h-1.98c-.003.046-.003.097-.003.147v.422h1.983v.427H3.93c.118.602.468 1.03 1.005 1.229a6.5 6.5 0 0 1 4.97-3.113A5.002 5.002 0 0 0 0 5zm16 5.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0zm-7.75 1.322c.069.835.746 1.485 1.964 1.562V14h.54v-.62c1.259-.086 1.996-.74 1.996-1.69 0-.865-.563-1.31-1.57-1.54l-.426-.1V8.374c.54.06.884.347.966.745h.948c-.07-.804-.779-1.433-1.914-1.502V7h-.54v.629c-1.076.103-1.808.732-1.808 1.622 0 .787.544 1.288 1.45 1.493l.358.085v1.78c-.554-.08-.92-.376-1.003-.787H8.25zm1.96-1.895c-.532-.12-.82-.364-.82-.732 0-.41.311-.719.824-.809v1.54h-.005zm.622 1.044c.645.145.943.38.943.796 0 .474-.37.8-1.02.86v-1.674l.077.018z"
              />
            </svg>
          </div>
          <h2
            class="title"
          >
            bitcoin
             
            (
            BTC
            )
             
            <br />
            <span>
              1233
               
              Views
            </span>
          </h2>
        </div>
        <div
          class="stats-separator dark-bg"
        >
          Stats for
           
          bitcoin
        </div>
        <div
          class="currency-details-list"
        >
          <div
            class="currency-details-row light-bg"
          >
            <div
              class="currency-details-left"
            >
              Rank:
            </div>
            <div
              class="currency-details-right"
            >
              1
            </div>
          </div>
          <div
            class="currency-details-row dark-bg"
          >
            <div
              class="currency-details-left"
            >
              Supply:
            </div>
            <div
              class="currency-details-right"
            >
              17 Million
            </div>
          </div>
          <div
            class="currency-details-row light-bg"
          >
            <div
              class="currency-details-left"
            >
              Max Supply:
            </div>
            <div
              class="currency-details-right"
            >
              21 Million
            </div>
          </div>
          <div
            class="currency-details-row dark-bg"
          >
            <div
              class="currency-details-left"
            >
              Market Cap (USD):
            </div>
            <div
              class="currency-details-right"
            >
              119 Billion
            </div>
          </div>
          <div
            class="currency-details-row light-bg"
          >
            <div
              class="currency-details-left"
            >
              Change (24hr):
            </div>
            <div
              class="currency-details-right"
            >
              -0.81
              %
               
              <svg
                class="negative-per"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.022 1.566a1.13 1.13 0 0 1 1.96 0l6.857 11.667c.457.778-.092 1.767-.98 1.767H1.144c-.889 0-1.437-.99-.98-1.767L7.022 1.566z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="currency-detail"
    >
      <a
        class="back-home-btn"
        href="/"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zm-4.5-.5a.5.5 0 0 1 0 1H5.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L5.707 7.5H11.5z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
      <div
        class="currency-title"
      >
        <div
          class="title-img"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 5a5.002 5.002 0 0 0 4.027 4.905 6.46 6.46 0 0 1 .544-2.073C3.695 7.536 3.132 6.864 3 5.91h-.5v-.426h.466V5.05c0-.046 0-.093.004-.135H2.5v-.427h.511C3.236 3.24 4.213 2.5 5.681 2.5c.316 0 .59.031.819.085v.733a3.46 3.46 0 0 0-.815-.082c-.919 0-1.538.466-1.734 1.252h1.917v.427h-1.98c-.003.046-.003.097-.003.147v.422h1.983v.427H3.93c.118.602.468 1.03 1.005 1.229a6.5 6.5 0 0 1 4.97-3.113A5.002 5.002 0 0 0 0 5zm16 5.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0zm-7.75 1.322c.069.835.746 1.485 1.964 1.562V14h.54v-.62c1.259-.086 1.996-.74 1.996-1.69 0-.865-.563-1.31-1.57-1.54l-.426-.1V8.374c.54.06.884.347.966.745h.948c-.07-.804-.779-1.433-1.914-1.502V7h-.54v.629c-1.076.103-1.808.732-1.808 1.622 0 .787.544 1.288 1.45 1.493l.358.085v1.78c-.554-.08-.92-.376-1.003-.787H8.25zm1.96-1.895c-.532-.12-.82-.364-.82-.732 0-.41.311-.719.824-.809v1.54h-.005zm.622 1.044c.645.145.943.38.943.796 0 .474-.37.8-1.02.86v-1.674l.077.018z"
            />
          </svg>
        </div>
        <h2
          class="title"
        >
          bitcoin
           
          (
          BTC
          )
           
          <br />
          <span>
            1233
             
            Views
          </span>
        </h2>
      </div>
      <div
        class="stats-separator dark-bg"
      >
        Stats for
         
        bitcoin
      </div>
      <div
        class="currency-details-list"
      >
        <div
          class="currency-details-row light-bg"
        >
          <div
            class="currency-details-left"
          >
            Rank:
          </div>
          <div
            class="currency-details-right"
          >
            1
          </div>
        </div>
        <div
          class="currency-details-row dark-bg"
        >
          <div
            class="currency-details-left"
          >
            Supply:
          </div>
          <div
            class="currency-details-right"
          >
            17 Million
          </div>
        </div>
        <div
          class="currency-details-row light-bg"
        >
          <div
            class="currency-details-left"
          >
            Max Supply:
          </div>
          <div
            class="currency-details-right"
          >
            21 Million
          </div>
        </div>
        <div
          class="currency-details-row dark-bg"
        >
          <div
            class="currency-details-left"
          >
            Market Cap (USD):
          </div>
          <div
            class="currency-details-right"
          >
            119 Billion
          </div>
        </div>
        <div
          class="currency-details-row light-bg"
        >
          <div
            class="currency-details-left"
          >
            Change (24hr):
          </div>
          <div
            class="currency-details-right"
          >
            -0.81
            %
             
            <svg
              class="negative-per"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.022 1.566a1.13 1.13 0 0 1 1.96 0l6.857 11.667c.457.778-.092 1.767-.98 1.767H1.144c-.889 0-1.437-.99-.98-1.767L7.022 1.566z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
